File: backend/internal/hub/models/repositories.go
package models

import (
	"time"

	"github.com/OrcaCD/orca-cd/internal/hub/crypto"
)

type RepositoryProvider string

const (
	GitHub      RepositoryProvider = "github"
	GitLab      RepositoryProvider = "gitlab"
	Generic     RepositoryProvider = "generic"
	Bitbucket   RepositoryProvider = "bitbucket"
	AzureDevOps RepositoryProvider = "azure_devops"
	Gitea       RepositoryProvider = "gitea"
)

type RepositoryAuthMethod string

const (
	AuthMethodNone  RepositoryAuthMethod = "none"
	AuthMethodToken RepositoryAuthMethod = "token"
	AuthMethodBasic RepositoryAuthMethod = "basic"
	AuthMethodSSH   RepositoryAuthMethod = "ssh"
)

type RepositorySyncType string

const (
	SyncTypePolling RepositorySyncType = "polling"
	SyncTypeWebhook RepositorySyncType = "webhook"
	SyncTypeManual  RepositorySyncType = "manual"
	// TODO add "scheduled" and maybe "github-app"
)

type RepositorySyncStatus string

const (
	SyncStatusUnknown RepositorySyncStatus = "unknown"
	SyncStatusSyncing RepositorySyncStatus = "syncing"
	SyncStatusFailed  RepositorySyncStatus = "failed"
	SyncStatusSuccess RepositorySyncStatus = "success"
)

type Repository struct {
	Base
	Name                              string                  `gorm:"type:text;not null"`
	Url                               string                  `gorm:"type:text;not null;uniqueIndex:idx_repositories_url_sync_type"`
	Provider                          RepositoryProvider      `gorm:"type:text;not null"`
	AuthMethod                        RepositoryAuthMethod    `gorm:"type:text;not null"`
	AuthUser                          *crypto.EncryptedString `gorm:"type:text;"`
	AuthToken                         *crypto.EncryptedString `gorm:"type:text;"`
	SyncType                          RepositorySyncType      `gorm:"type:text;not null;uniqueIndex:idx_repositories_url_sync_type"`
	SyncStatus                        RepositorySyncStatus    `gorm:"type:text;not null"`
	LastSyncError                     *string                 `gorm:"type:text;"`
	PollingInterval                   *time.Duration          `gorm:"type:integer;"`
	WebhookSecret                     *crypto.EncryptedString `gorm:"type:text;"`
	LastSyncedAt                      *time.Time              `gorm:"type:timestamp;"`
	CreatedBy                         string                  `gorm:"type:text;not null;"`
	GitHubActionsOIDCEnabled          bool                    `gorm:"column:github_actions_oidc_enabled;not null;default:false"`
	GitHubActionsOIDCAllowRepoSync    bool                    `gorm:"column:github_actions_oidc_allow_repo_sync;not null;default:true"`
	GitHubActionsOIDCAllowImageSync   bool                    `gorm:"column:github_actions_oidc_allow_image_sync;not null;default:true"`
	GitHubActionsOIDCAllowedBranches  []string                `gorm:"column:github_actions_oidc_allowed_branches;type:text;serializer:json;not null"`
	GitHubActionsOIDCAllowedWorkflows []string                `gorm:"column:github_actions_oidc_allowed_workflows;type:text;serializer:json;not null"`
	Applications                      []Application           `gorm:"foreignKey:RepositoryId;"`
}

func (Repository) TableName() string {
	return "repositories"
}

File: frontend/src/lib/repositories.ts
import { fetcher } from "./api";

export type RepositoryProvider =
	| "github"
	| "gitlab"
	| "generic"
	| "bitbucket"
	| "azure_devops"
	| "gitea";

export type RepositoryAuthMethod = "none" | "token" | "basic" | "ssh";
export type RepositorySyncType = "polling" | "webhook" | "manual";
export type RepositorySyncStatus = "unknown" | "syncing" | "failed" | "success";

export interface Repository {
	id: string;
	name: string;
	url: string;
	provider: RepositoryProvider;
	authMethod: RepositoryAuthMethod;
	syncType: RepositorySyncType;
	syncStatus: RepositorySyncStatus;
	lastSyncError: string | null;
	pollingIntervalSeconds: number | null;
	lastSyncedAt: string | null;
	createdBy: string;
	createdAt: string;
	updatedAt: string;
	appCount: number;
	webhookSecret?: string | undefined;
	webhookUrl?: string | undefined;
}

export interface CreateRepositoryRequest {
	url: string;
	provider: RepositoryProvider;
	authMethod: RepositoryAuthMethod;
	authUser?: string;
	authToken?: string;
	syncType: RepositorySyncType;
	pollingIntervalSeconds?: number;
	webhookSecret?: string;
}

export interface UpdateRepositoryRequest {
	authMethod?: RepositoryAuthMethod;
	authUser?: string;
	authToken?: string;
	syncType?: RepositorySyncType;
	pollingIntervalSeconds?: number;
	webhookSecret?: string;
}

export interface TestConnectionRequest {
	url: string;
	provider: RepositoryProvider;
	authMethod: RepositoryAuthMethod;
	authUser?: string;
	authToken?: string;
}

export function createRepository(data: CreateRepositoryRequest): Promise<Repository> {
	return fetcher<Repository>("/repositories", "POST", data);
}

export function testRepositoryConnection(data: TestConnectionRequest): Promise<void> {
	return fetcher("/repositories/test-connection", "POST", data);
}

export function deleteRepository(id: string): Promise<void> {
	return fetcher(`/repositories/${id}`, "DELETE");
}

export function updateRepository(id: string, data: UpdateRepositoryRequest): Promise<Repository> {
	return fetcher<Repository>(`/repositories/${id}`, "PUT", data);
}

export function syncRepository(id: string): Promise<void> {
	return fetcher(`/repositories/${id}/sync`, "POST");
}

export function getGitProviderIconPath(provider: RepositoryProvider): string {
	const iconName = provider === "generic" ? "git" : provider;

	return `/assets/icons/${iconName}.svg`;
}

export function getGitProviderIconClass(provider: RepositoryProvider): string {
	return provider === "github" ? "dark:invert" : "";
}
